<!doctype html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="description" content="Science, meet productivity, A computational RNA biologist exploring productivity, python, and reproducibility.">

        <link rel="alternate"  href="/feeds/all.atom.rss.xml" type="application/atom+xml" title="Science, meet productivity Full Atom Feed"/>
        <link rel="alternate" href="/feeds/all.rss.xml" type="application/rss+xml" title="Science, meet productivity Full RSS Feed"/>

        <title>CodeNeuro NYCÂ 2015 // Science, meet productivity // A computational RNA biologist exploring productivity, python, and reproducibility.</title>

    <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/pure/0.3.0/pure-min.css">
    <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/font-awesome/4.1.0/css/font-awesome.min.css">
    <link rel="stylesheet" href="../../../../../theme/css/pure.css">
    <link rel="stylesheet" href="../../../../../theme/css/pygments.css">

    <script src="//cdnjs.cloudflare.com/ajax/libs/jquery/2.0.3/jquery.min.js"></script>
    <script src="//cdnjs.cloudflare.com/ajax/libs/fitvids/1.0.1/jquery.fitvids.min.js"></script>
    <script>
        $(document).ready(function(){
            $(".content").fitVids();
        });
    </script>
</head>

<body>
    <div class="pure-g-r" id="layout">
        <div class="sidebar pure-u">
            <div class="cover-img" style="background-image: url('https://raw.githubusercontent.com/olgabot/olgabot.github.io-source/master/content/images/ordered_driftwood.jpg')">
                <div class="cover-body">
                    <header class="header">
                        <hgroup>
                            <img class="avatar" src="http://raw.githubusercontent.com/olgabot/olgabot.github.io-source/master/content/images/olga_icon_square.jpg">
                            <h1 class="brand-main"><a href="/">Science, meet productivity</a></h1>
                            <p class="tagline">A computational RNA biologist exploring productivity, python, and reproducibility.</p>
                                <p class="links"><a href="../../../../../archives.html">Archive</a></p>
                                <p class="links"><a href="../../../../../pages/about.html">About</a></p>
                                <p class="social">
                                    <a href="http://github.com/olgabot">
                                        <i class="fa fa-github-square fa-3x"></i>
                                    </a>
                                    <a href="http://twitter.com/olgabot">
                                        <i class="fa fa-twitter-square fa-3x"></i>
                                    </a>
                                    <a href="http://www.linkedin.com/in/olgabotvinnik/">
                                        <i class="fa fa-linkedin-square fa-3x"></i>
                                    </a>
                                <p>
                        </hgroup>
                    </header>
                </div>
            </div>
        </div>
    <div class="pure-u">
        <div class="content">
            <section class="post">
                <header class="post-header">
                    <h1>CodeNeuro <span class="caps">NYC</span>&nbsp;2015</h1>
                        <p class="post-meta">
                            // under                                 <a class="post-category" href="../../../../../tag/programming.html">programming</a>
                                <a class="post-category" href="../../../../../tag/conferences.html">conferences</a>
                                <a class="post-category" href="../../../../../tag/bioinformatics.html">bioinformatics</a>
                                <a class="post-category" href="../../../../../tag/neuroscience.html">neuroscience</a>
                        </p>
                </header>
            </section>
            <p>Last weekend I attended the <a href="http://codeneuro.org/2015/NYC/">CodeNeuro <span class="caps">NYC</span></a>
conference, which is a conference for neuroscientists, programmers, and
everything in between. It was a unique experience because it had both the
excitement and energy of a tech conference, and the pure, unadulterated zest
for knowledge of a scientific&nbsp;conference.</p>
<h2>Day 1: Talks and&nbsp;panel</h2>
<p>The first day of the conference started after the workday on Friday at 5pm at
the <a href="http://www.newmuseum.org/">New Museum</a> in Little Italy, Manhattan, <span class="caps">NYC</span>.
This was the first time I&#8217;d been to this place and it was really awesome, even
from the outside! Check it out, plus some other photos I took from the&nbsp;conference:</p>
<blockquote class="imgur-embed-pub" lang="en" data-id="a/q3c57"><a href="//imgur.com/a/q3c57">codeneuro nyc 2015</a></blockquote>

<script async src="//s.imgur.com/min/embed.js" charset="utf-8"></script>

<p>(<span class="caps">FYI</span> If you want to look through all the tweets from the entire event, check out this <a href="https://storify.com/yangbodu/codeneuro">Storify</a>&nbsp;board.)</p>
<p>The conference kicked off with some mingling and beer. I gravitated towards the three I knew people coming in to the conference, <a href="http://www.janelia.org/people/scientist/nick-sofroniew">Nick
Sofroniew</a>, with whom
I did an <a href="http://www.janelia.org/student-programs/undergraduate-program#722">undergrad summer program</a> and invited me to talk, <a href="http://www.jeremyfreeman.net/">Jeremy Freeman</a>, the main organizer of
the whole freakin&#8217; thing, and <a href="https://twitter.com/bensussman">Ben
Sussman</a>, who co-taught the <code>gitgoing</code>
tutorial with me and is an all-around awesome person. But the space was rad,
we had a balcony and a nice view of the city, and I got to meet a bunch of
people doing super interesting things in&nbsp;neuroscience.</p>
<blockquote class="twitter-tweet" lang="en"><p>view from <a href="https://twitter.com/CodeNeuro">@CodeNeuro</a>&#8230; <a href="http://t.co/kInTbaFWHW">pic.twitter.com/kInTbaFWHW</a></p>&mdash; Angela Radulescu (@angitweets) <a href="https://twitter.com/angitweets/status/586659812198522880">April 10, 2015</a></blockquote>

<script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script>

<p>We got started with the talks, which we had three 10 minute (ish) talks, followed by a half hour (ish) break. It was a nice format that we could chat and break in between the talks not only to get a mental break but also discuss the concepts presented by the previous talks in the context of neuroscience. Here&#8217;s all the&nbsp;talks:</p>
<ol>
<li>Adam Packer (<span class="caps">UCL</span>): All optical recording and manipulation of individual&nbsp;neurons</li>
<li>Eiman Azim (Columbia): Quantitative behavior and mouse genetics for probing motor circuit&nbsp;function</li>
<li>Randal Burns (Johns Hopkins): Open Connectome project using <span class="caps">CATMAID</span> [<a href="https://github.com/openconnectome/ocpcatmaid">github repo</a>]</li>
<li>Paco Nathan (Databricks): Spark for big&nbsp;data</li>
<li>Eliza Chang (The Data Incubator): PhD to Data&nbsp;Scientist</li>
<li>Michael Dewar (<span class="caps">NYT</span>): Real-time analytics on streaming data [<a href="https://github.com/mikedewar/codeneuro_streamtools">github repo for presentation</a>] | [<a href="https://github.com/nytlabs/streamtools">github repo for <code>streamtools</code> package</a>] </li>
<li>Hilary Parker (Etsy): Data science at Etsy (+cute statistical&nbsp;plushies!)</li>
<li>Brendan Lake (<span class="caps">NYU</span>): Teaching computers to scribble characters like&nbsp;humans</li>
<li>Olga Botvinnik, me! (<span class="caps">UCSD</span>): <code>flotilla</code>: An open-source Python package for iterative machine learning analyses [<a href="http://nbviewer.ipython.org/format/slides/gist/olgabot/2ee1087d74df46c842df#/">slides</a>]</li>
</ol>
<p>As you can imagine from the caliber of all the other talks, I was super nervous for mine! because I&#8217;m not doing things at exabyte scale, I&#8217;m not doing streaming data, and I&#8217;m not even doing neuroscience! I&#8217;m just analyzing static datasets with the usual machine learning algorithms. But as Jeremy pointed out, we need better tools for analyzing the existing non-big data datasets that we have, plus I got great questions and feedback from the talk which was&nbsp;great.</p>
<p>My top three favorite talks&nbsp;were:</p>
<p><strong><span class="caps">NYT</span> R&amp;D demo</strong> This was a fantastic talk about streaming data analytics.  He
and his team had created an interactive way to analyze real-time data. They
had a <span class="caps">GUI</span> interface where you could select boxes that represented an action on
data (<a href="http://en.wikipedia.org/wiki/Extract,_transform,_load">extract, transform, load aka
<span class="caps">ETL</span></a>-like things), and
you could string these boxes together to create an analysis pipeline. It
reminded me a lot of the <a href="https://usegalaxy.org/">Galaxy</a> interface for
reproducible bioinformatics pipelines. Check out the tweet below for a&nbsp;screenshot:</p>
<blockquote class="twitter-tweet" lang="en"><p>Picture of <a href="https://twitter.com/hashtag/realTimeData?src=hash">#realTimeData</a> including mine <a href="https://twitter.com/hashtag/codeNeuro?src=hash">#codeNeuro</a> <a href="http://t.co/VCIfFlFHTK">pic.twitter.com/VCIfFlFHTK</a></p>&mdash; aybuke turker (@aybuketurker) <a href="https://twitter.com/aybuketurker/status/586678871942111232">April 10, 2015</a></blockquote>

<script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script>

<p><strong>Optogenetics</strong>.</p>
<blockquote class="twitter-tweet" lang="en"><p>A. Packer shows some amazing tools to simultaneously record + stimulate ensembles of neurons with <a href="https://twitter.com/hashtag/CaImaging?src=hash">#CaImaging</a> and <a href="https://twitter.com/hashtag/optogenetics?src=hash">#optogenetics</a>. <a href="https://twitter.com/hashtag/codeneuro?src=hash">#codeneuro</a></p>&mdash; Felipe Gerhard (@neuroflips) <a href="https://twitter.com/neuroflips/status/586685623068717056">April 11, 2015</a></blockquote>

<script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script>

<p><strong>Spark</strong>. This was an awesome talk by <a href="http://liber118.com/pxn/">Paco Nathan</a>
on using big data tools. Paco had worked with both Hadoop and Spark and gave a
great overview of the history of Big Data (which he pinpointed to Q3 1997),
the different papers about it and tools for it, and where Spark fits in that
ecosystem. He gave a few use cases and was overall just a very entertaining&nbsp;presenter.</p>
<blockquote class="twitter-tweet" lang="en"><p>Great talk by <a href="https://twitter.com/placoid">@placoid</a> at <a href="https://twitter.com/CodeNeuro">@CodeNeuro</a> just now - apparently rotational bearings on planes create 12 exabytes of data / day&#8230; wow <a href="https://twitter.com/hashtag/bigdata?src=hash">#bigdata</a></p>&mdash; Alice Lloyd George (@<span class="caps">AMLG23</span>) <a href="https://twitter.com/AMLG23/status/586689726196883458">April 11, 2015</a></blockquote>

<script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script>

<p>[The most retweeted talks?? Maybe a Storify of the @CodeNeuro account for April 10th&nbsp;2015]</p>
<p>After all the talks, there was a panel led by Jeremy with some giants of
neuroscientists, Tony Movshon, Eve Marder and Larry Abbott. A recurring theme
was that the fundamental questions of how the brain works hasn&#8217;t changed, but
the tools have drastically&nbsp;changed.</p>
<p>One story that stuck with me was Eve describing that when she was doing
experiments, she used an
<a href="http://en.wikipedia.org/wiki/Oscilloscope">oscilloscope</a> to listen (yes, with
her ears! I thought this was amazing!) to the measurements from neurons, and
there was a direct, phsyical, visceral reaction that when you designed an
experiment and heard specific tones, you knew <strong>exactly</strong> what that meant,
because you were intimately familiar with the exact tones and sounds produced
by the device. But now, she said, the high throughput methods, while capable
of measuring many things at once, are so detached from the physicality of
neurons firing that she feels people aren&#8217;t connected to what they&#8217;re
measuring anymore. I found that interesting because in my field of <span class="caps">RNA</span>
biology, we&#8217;ll do these newfangled <a href="http://en.wikipedia.org/wiki/DNA_sequencing#Next-
generation_methods">high-throughput sequencing
experiments</a> to measure <span class="caps">RNA</span> expression as a proxy for protein
abundance, and study the <a href="http://en.wikipedia.org/wiki/Alternative_splicing">alternative
splicing</a> of these <span class="caps">RNA</span>
transcripts, but every time, we always have to validate our findings using an
older, low-throughput method like <a href="http://en.wikipedia.org/wiki/Revers
e_transcription_polymerase_chain_reaction"><span class="caps">RT</span>-<span class="caps">PCR</span></a>. I don&#8217;t know what the equivalent
experiments are in neuroscience, because it seems all very magical to me that
they&#8217;re able to manipulate individual neurons in mice, but I also wonder if
it&#8217;s known that the neuron in the exact same position in one animal and
another animal, do the exact same thing. So far, biology has relied on laws of
averages, so if you do a high throughput experiment on a bunch of cells and
get an average signal, you should get that same result if you do a low-
throughput version of that experiment. I don&#8217;t know what the equivalent of
that is in neuroscience, but it was refreshing to hear that neuro has its own
share of growing pains as its adopted its own high-throughput techniques, and
that biology wasn&#8217;t the only&nbsp;one.</p>
<p>It was interesting to hear about the challenges of neuroscience, of how
excited people are about optogenetics, and how many people are apparently
skeptical of <a href="http://en.wikipedia.org/wiki/Functional_magnetic_resonance_imaging">functional magentic resonance
imaging</a>
(fMRI) data, and how nobody knows what the <a href="http://en.wikipedia.org/wiki/Cerebral_cortex">cerebral
cortex</a>&nbsp;does!</p>
<h2>Day 2: Tutorials +&nbsp;Hackathon</h2>
<p>The next morning we were in the <a href="http://www.newinc.org/">New, Inc</a> coworking
space, and started off with some kickass bagels and&nbsp;lox/salmon:</p>
<blockquote class="twitter-tweet" lang="en"><p>breakfast begins <a href="http://t.co/55QljXZqkw">pic.twitter.com/55QljXZqkw</a></p>&mdash; CodeNeuro (@CodeNeuro) <a href="https://twitter.com/CodeNeuro/status/586893786212651008">April 11, 2015</a></blockquote>

<script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script>

<p>Then, people split up about 50/50 for the tutorials and the&nbsp;hackathon.</p>
<h3><code>gitgoing</code> tutorial</h3>
<p>Then, Ben Sussman and I taught a tutorial called <a href="https://github.com/CodeNeuro/gitgoing">gitgoing</a> to quickly teach scientists the version control system <a href="http://en.wikipedia.org/wiki/Git_%28software%29"><code>git</code></a> and code testing via <a href="http://pytest.org/"><code>py.test</code></a> and integration systems via <a href="https://travis-ci.org/">Travis-<span class="caps">CI</span></a>. We had about 20-30 attendees. The goal was to get the scientists acquainted with common tools in open source software so that they could contribute themselves. It was kind of like a mini <a href="http://software-carpentry.org/">Software Carpentry</a> workshop, but we assumed our target audience had some coding experience, and we didn&#8217;t take the time to explain computer science fundamentals like variables, loops, flow control,&nbsp;etc.</p>
<p>Our class was structured exactly as laid out in the <code>README.md</code> file. First we setup their computers so they had <code>git</code> and Python 2.7. This took about an hour total to get everyone done. Some people finished faster and started moving on to the <code>git</code> section. Then Ben did an awesome explanation of <code>git</code>, and I learned a bunch of stuff! I didn&#8217;t realize that when you <code>git clone</code> a repo, you&#8217;re getting <span class="caps">ENTIRE</span> history of the project, so that makes sense why downloading all of <a href="ipython.org">IPython</a>/<a href="https://jupyter.org/">Jupyter</a> takes forever. It was also a really helpful analogy to describe the entire repository as an &#8220;ocean of code,&#8221; and that a branch is a single window into that ocean. We also talked about merge conflicts, and how they can be really easy to create if, say someone renamed one of the arguments of a function, and someone else added an argument, and then <code>git</code> doesn&#8217;t know what to do anymore. They picked up on testing pretty quickly, and someone asked &#8220;well if <code>git</code> thinks it&#8217;s okay, but how do you know the code will run?&#8221; Which brought us directly to&nbsp;testing!</p>
<p>Next, I talked about testing and why it&#8217;s important. I wrote some simple Python code with functions like <code>mean_plus_one</code>, <code>std_plus_one</code>, and <code>cv</code> (coefficient of variance). They were just slight variations on the true <a href="http://www.numpy.org/"><code>numpy</code></a> functions so the learners couldn&#8217;t just use the <code>numpy</code> version. We looked at the test file, <code>test_gitgoing.py</code> which used <code>py.test</code><span class="quo">&#8216;</span>s fixtures, which take care of the <code>setUp</code> and <code>tearDown</code> methods that some other testing frameworks have. We saw a simple example of fixtures, which creates a 20x10 matrix of normally distributed random numbers. These could have been set as integers, I just wanted to illustrate how you can create new fixtures from existing&nbsp;ones.</p>
<div class="highlight"><pre><span class="kn">import</span> <span class="nn">numpy</span> <span class="kn">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pytest</span>

<span class="nd">@pytest.fixture</span>
<span class="k">def</span> <span class="nf">n_rows</span><span class="p">():</span>
    <span class="k">return</span> <span class="mi">20</span>

<span class="nd">@pytest.fixture</span>
<span class="k">def</span> <span class="nf">n_cols</span><span class="p">():</span>
    <span class="k">return</span> <span class="mi">10</span>

<span class="nd">@pytest.fixture</span>
<span class="k">def</span> <span class="nf">x_norm</span><span class="p">(</span><span class="n">n_rows</span><span class="p">,</span> <span class="n">n_cols</span><span class="p">):</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randn</span><span class="p">(</span><span class="n">n_rows</span><span class="p">,</span> <span class="n">n_cols</span><span class="p">)</span>
</pre></div>


<p>There is a commented out broken test in the repo which they had to fix, which was a great formative assessment because they had to use their newly formed <code>git</code> and testing knowledge to fix the test, commit the changes, make a branch of their &#8220;feature&#8221; fixing the test, pushing to the branch, and making a pull request to the master branch. It was really rewarding for the students to see their pull request on Github, and to see their commits on the network of contributors to the <code>gitgoing</code> repo.</p>
<p>We unfortunately didn&#8217;t advertise the #gitgoing hashtag before the tutorial so we didn&#8217;t get any live-tweets, but Ben and I mitigated this and took a picture&nbsp;afterwards:</p>
<blockquote class="twitter-tweet" lang="en"><p><a href="https://twitter.com/hashtag/gitgoing?src=hash">#gitgoing</a> at <a href="https://twitter.com/CodeNeuro">@CodeNeuro</a>! Thanks <a href="https://twitter.com/bensussman">@bensussman</a> for being an awesome co-teacher and everyone for coming! <a href="http://t.co/l0BUhoeWja">pic.twitter.com/l0BUhoeWja</a></p>&mdash; Olga Botvinnik (@olgabot) <a href="https://twitter.com/olgabot/status/586980895397019650">April 11, 2015</a></blockquote>

<script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script>

<h3>Spark&nbsp;tutorial</h3>
<p>After the <code>gitgoing</code> tutorial we had a break and mingled. Then, I went to the Spark tutorial, taught by <a href="">Paco Nathan</a> from <a href="">DataBricks</a> (slides <a href="http://training.databricks.com/workshop/dbc_intro.pdf">here</a>). I&#8217;ve seen Spark demos before but I haven&#8217;t put in the time to play around with the tools, so this was a great way to get&nbsp;exposed!</p>
<p>I was a little late to the tutorial, so I missed the initial setup. I was handed a slip of paper with a url, username and password that was my personal login to the DataBricks cloud. Paco was doing a &#8220;preflight check&#8221; of explaining different Spark concepts before we dove in. The key things I took away&nbsp;were:</p>
<ul>
<li><strong><span class="caps">RDD</span></strong>: Resilient Distributed Dataset. This is the core unit of a spark analysis, where you load in data, and use Spark to indicate that you want it to be&nbsp;parallelized.</li>
<li><strong><code>sc</code></strong>: &#8220;Spark context.&#8221; This is the object that exists in all the Python library versions with Spark, and is the object that you will be using to create and operate on&nbsp;datasets. </li>
</ul>
<p>We used an IPython notebook-style <a href="http://en.wikipedia.org/wiki/Read%E2%80%93eval%E2%80%93print_loop"><span class="caps">REPL</span></a> interface, which was really nice because we could see the output from our commands right away. We continued with learning how&nbsp;to:</p>
<ol>
<li>Read text files and separate lines by&nbsp;tabs</li>
<li>Filter for the first word of the line to be <code>ERROR</code></li>
<li>Get all the second words of the error&nbsp;lines</li>
<li>Find the errors that were caused by <code>"mysql"</code>.</li>
</ol>
<p>You can see a summary of my notes&nbsp;here:</p>
<div class="highlight"><pre>lines = sc.textFile(&quot;/mnt/paco/intro/error_log.txt&quot;) \
  .map(lambda x: x.split(&quot;\t&quot;))
errors = lines.filter(lambda x: x[0] == &quot;ERROR&quot;)
messages = errors.map(lambda x: x[1])
messages.filter(lambda x: x.find(&quot;mysql&quot;) &gt; -1).count()

# At this point, the data is loaded into memory on the workers and you don&#39;t
# need to attack disk again, so this operation is very fast
messages.filter(lambda x: x.find(&quot;php&quot;) &gt; -1).count()
</pre></div>


<p>Yay, we learned how to read files! Always good&nbsp;:)</p>
<p>The next part of the tutorial, we learned how to use <code>flatMap</code>, and then how to <code>join</code> several RDDs on their <code>(key, value)</code> pairs. After the break, we had a mini lecture about &#8220;Computational thinking&#8221; where we had to use what we learned so far to find all the instances of the word &#8220;Spark&#8221; in two files. If you&#8217;re interested, you can see my full notes <a href="https://gist.github.com/olgabot/ab058876b3bda6198f25">here</a>.</p>
<h3>Hackathon</h3>
<p>The hackathon was centered around the <a href="http://codeneuro.org/neurofinder/">neurofinder</a> challenge, which is to try and extract neuronal signals from Calcium imaging data. The goals were&nbsp;go:</p>
<ol>
<li>Refine a standardized <span class="caps">API</span> for &#8220;source extraction&#8221; algorithms, including input/ouptu&nbsp;formats</li>
<li>Work on evaluation metrics for algorithms and agree on ground truth&nbsp;definition</li>
<li>Work on incorporating existing algorithms into this <span class="caps">API</span></li>
<li>Work on front end/back end of a website that would automatically run submitted algorithms on the test data, get the results and upload them to a&nbsp;leaderboard</li>
</ol>
<p>First, everyone introduced themselves and the group organically split into a few groups: the website, <span class="caps">API</span> input/output formatting, implementing algorithms, and designing&nbsp;metrics.</p>
<p>The website team was composed of five people who had web development frontend/backend experience went to work on the website. They generated some prototype websites and code <strong>Jeremy/Nick: is there a link for the code&nbsp;here??</strong></p>
<p>Another group of about 25 worked on the <span class="caps">API</span> and the input/output formats, with their final notes added to <a href="https://github.com/CodeNeuro/neurofinder/wiki">this wiki</a>.</p>
<p>Another group of 10-20 split to work on incorporating existing algorithms in to this <span class="caps">API</span>. The rest worked on defining evaluation metrics for algorithms and the &#8220;ground truth&#8221; definitions of &#8220;what is a neuron&#8221; for these methods. They implemented some evaluation metrics, and developed an initial ground truth definition of manual centers an morphological boundaries. While this initial &#8220;ground truth&#8221; may be circular, it&#8217;s a great&nbsp;start.</p>
<p>There were still people working well into the&nbsp;evening!</p>
<blockquote class="twitter-tweet" lang="en"><p>the codeneuro hardcore, still rocking! <a href="http://t.co/TyxmlSNOZB">pic.twitter.com/TyxmlSNOZB</a></p>&mdash; CodeNeuro (@CodeNeuro) <a href="https://twitter.com/CodeNeuro/status/587072775057231872">April 12, 2015</a></blockquote>

<script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script>

<p>Afterwards, we all hung out at a nice&nbsp;dinner:</p>
<p><img alt="CodeNeuro after-dinner party" src="http://i.imgur.com/KY9smQR.jpg" /></p>
<h2>Acknowledgements</h2>
<p>Finally, I want to acknowledge all the organizers of the&nbsp;event:</p>
<ul>
<li>Jeremy&nbsp;Freeman</li>
<li>Nick&nbsp;Sofroniew</li>
<li><a href="http://graphics.stanford.edu/~broxton/">Michael&nbsp;Broxton</a></li>
<li><a href="http://web.stanford.edu/~logang/">Logan&nbsp;Grosenick</a></li>
<li><a href="https://twitter.com/dgangul1">Deep&nbsp;Ganguli</a></li>
<li><a href="https://twitter.com/hackingdata">Jeff&nbsp;Hammerbacher</a></li>
</ul>
            <a href="#" class="go-top">Go Top</a>
    <div class="comments">
        <div id="disqus_thread"></div>
        <script type="text/javascript">
            /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
            var disqus_shortname = "sciencemeetproductivity"; // required: replace example with your forum shortname

            /* * * DON'T EDIT BELOW THIS LINE * * */
            (function() {
                var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
                dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
                (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
            })();
        </script>
        <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
        <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
    </div>
<footer class="footer">
    <p>&copy; Olga Botvinnik &ndash;
        Built with <a href="https://github.com/PurePelicanTheme/pure-single">Pure Theme</a>
        for <a href="http://blog.getpelican.com/">Pelican</a>
    </p>
</footer>        </div>
    </div>
    </div>
    <script>
        var $top = $('.go-top');

        // Show or hide the sticky footer button
        $(window).scroll(function() {
            if ($(this).scrollTop() > 200) {
                $top.fadeIn(200);
            } else {
                $top.fadeOut(200);
            }
        });

        // Animate the scroll to top
        $top.click(function(event) {
            event.preventDefault();
            $('html, body').animate({scrollTop: 0}, 300);
        })

        // Makes sure that the href="#" attached to the <a> elements
        // don't scroll you back up the page.
        $('body').on('click', 'a[href="#"]', function(event) {
            event.preventDefault();
        });
    </script>
    <script type="text/javascript">
        var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
        document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
    </script>
    <script type="text/javascript">
        try {
            var pageTracker = _gat._getTracker("UA-53680167-1");
            pageTracker._trackPageview();
            } catch(err) {}
    </script>

</body>
</html>