<!doctype html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="description" content="Science, meet productivity, A computational RNA biologist exploring productivity, python, and reproducibility.">


        <title>Implementing annotation-free quantification of splice junction usage via RNA-STAR and Python+PandasÂ magic // Science, meet productivity // A computational RNA biologist exploring productivity, python, and reproducibility.</title>

    <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/pure/0.3.0/pure-min.css">
    <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/font-awesome/4.1.0/css/font-awesome.min.css">
    <link rel="stylesheet" href="../../../../../theme/css/pure.css">
    <link rel="stylesheet" href="../../../../../theme/css/pygments.css">

    <script src="//cdnjs.cloudflare.com/ajax/libs/jquery/2.0.3/jquery.min.js"></script>
    <script src="//cdnjs.cloudflare.com/ajax/libs/fitvids/1.0.1/jquery.fitvids.min.js"></script>
    <script>
        $(document).ready(function(){
            $(".content").fitVids();
        });
    </script>
</head>

<body>
    <div class="pure-g-r" id="layout">
        <div class="sidebar pure-u">
            <div class="cover-img" style="background-image: url('https://raw.githubusercontent.com/olgabot/olgabot.github.io-source/master/content/images/waves.jpg')">
                <div class="cover-body">
                    <header class="header">
                        <hgroup>
                            <img class="avatar" src="http://raw.githubusercontent.com/olgabot/olgabot.github.io-source/master/content/images/olga_icon_square.jpg">
                            <h1 class="brand-main"><a href="/">Science, meet productivity</a></h1>
                            <p class="tagline">A computational RNA biologist exploring productivity, python, and reproducibility.</p>
                                <p class="social">
                                    <a href="http://github.com/olgabot">
                                        <i class="fa fa-github-square fa-3x"></i>
                                    </a>
                                    <a href="http://twitter.com/olgabot">
                                        <i class="fa fa-twitter-square fa-3x"></i>
                                    </a>
                                    <a href="http://www.linkedin.com/in/olgabotvinnik/">
                                        <i class="fa fa-linkedin-square fa-3x"></i>
                                    </a>
                                <p>
                        </hgroup>
                    </header>
                </div>
            </div>
        </div>
    <div class="pure-u">
        <div class="content">
            <section class="post">
                <header class="post-header">
                    <h1>Implementing annotation-free quantification of splice junction usage via <span class="caps">RNA</span>-<span class="caps">STAR</span> and Python+Pandas&nbsp;magic</h1>
                        <p class="post-meta">
                            // under                                 <a class="post-category" href="../../../../../tag/rna-splicing.html">rna splicing</a>
                                <a class="post-category" href="../../../../../tag/alternative-splicing.html">alternative splicing</a>
                                <a class="post-category" href="../../../../../tag/python.html">python</a>
                                <a class="post-category" href="../../../../../tag/pandas.html">pandas</a>
                                <a class="post-category" href="../../../../../tag/rna.html">rna</a>
                                <a class="post-category" href="../../../../../tag/science.html">science</a>
                        </p>
                </header>
            </section>
            <p>If you&#8217;ve been reading my blog for general python/productivity stuff, you are in for a <span class="caps">TREAT</span>! Here is some <span class="caps">REAL</span> <span class="caps">SCIENCE</span>! I&#8217;ll give a quick intro but please let me know if anything is unclear. I spend most of my time talking to a few people who study this stuff in painful detail, so I&#8217;d love to know where my explanation becomes&nbsp;cloudy.</p>
<p><strong><em>tl;dr: I wrote a small tool for quantifying alternative junction usage from <span class="caps">RNA</span>-<span class="caps">STAR</span> alignment output: https://github.com/olgabot/sj2psi And it&#8217;s on PyPI so you can <code>pip install sj2psi</code> as&nbsp;well.</em></strong></p>
<p>Alternative splicing is a method of creating different versions of a gene using the same <span class="caps">DNA</span>. It is awesome because it increases the possible number of genes in say the human genome from just 30,000 to millions. An extreme example is the <em>Drosoephila</em> (fruit fly) gene <em>Dscam</em>, which has over 60,000 potential versions, more than the genes in the <em>Drosophila</em> genome! Alternative splicing happens by combining different &#8220;<em>exons</em>&#8221; together (always in linear order) to create different proteins (or other gene products like <a href="http://en.wikipedia.org/wiki/Non-coding_RNA">ncRNAs</a> but we won&#8217;t go into&nbsp;that)</p>
<p><img alt="" src="http://upload.wikimedia.org/wikipedia/commons/0/0a/DNA_alternative_splicing.gif" /></p>
<p>These different versions of the same gene are called &#8220;isoforms.&#8221; What we care about is looking at individual events because we study <a href="http://en.wikipedia.org/wiki/RNA-binding_protein"><span class="caps">RNA</span> binding proteins</a> (RBPs), some of which have an effect on alternative splicing when they bind <span class="caps">RNA</span>. So maybe we know that an <span class="caps">RBP</span> binds some gene&#8217;s <span class="caps">RNA</span>, but it (most likely) only has an effect on the exon(s) immediately before or after. For example, the <span class="caps">RBP</span> <span class="caps">FOX2</span> is one of <a href="http://yeolab.ucsd.edu/yeolab/Home.html">our lab&#8217;s</a> favorites, partly because our boss found that if <span class="caps">FOX2</span> binds upstream (&#8220;before&#8221; in <span class="caps">DNA</span> terms) of an exon, it suppresses inclusion of it. <span class="caps">HOWEVER</span>! <span class="caps">FOX2</span> binding downstream (&#8220;after&#8221;) an exon promotes inclusion! Check out this figure from <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2735254/">Yeo et al, <em>Nat Struct Mol Bio</em> (2009)</a>:</p>
<p><img alt="" src="https://31.media.tumblr.com/8749e1c26cb36237e1d8d2ea874327f1/tumblr_inline_n4wpkodHzT1rw6gvj.png" /></p>
<p>Quantifying alternative exon usage and changes is always a challenge, and this is especially difficult for unannotated splicing events. Which is why I was excited when I read <a href="http://genome.cshlp.org/content/early/2013/12/03/gr.161034.113?top=1">a single-cell paper</a> learned of <a href="https://github.com/pervouchine/bam2ssj"><code>bam2ssj</code></a> (Paper: <a href="http://www.ncbi.nlm.nih.gov/pubmed/23172860">Pervouchine et al, <em>Bioinformatics</em> (2013)</a> and its successor, <a href="https://github.com/pervouchine/sjcount"><code>sjcount</code></a>. The promise is that instead of quantifying whole splicing events which have to fall into some specific category (<span class="caps">SE</span>/<span class="caps">MXE</span>/<span class="caps">RI</span>/<span class="caps">A5SS</span>/<span class="caps">A3SS</span>/<span class="caps">AFE</span>/<span class="caps">ALE</span>/TandemUTR, see this figure from <a href="http://www.nature.com/nature/journal/v456/n7221/abs/nature07509.html">Wang <span class="caps">ET</span> et al, <em>Nature</em> (2008)</a>. Each &#8220;box&#8221; is an exon and the lines in between exons are stretches of <span class="caps">DNA</span> that are removed (called &#8220;introns&#8221; because they &#8220;interrupt&#8221; the&nbsp;exons)</p>
<p><img alt="" src="http://www.nature.com/nature/journal/v456/n7221/images/nature07509-f2.2.jpg" /></p>
<p>Plus these events have to be annotated for use with my splicing quantifier of choice (<a href="http://miso.readthedocs.org/en/fastmiso/"><span class="caps">MISO</span></a>), or you could try to pull them out from the aligned file. However, I work with single-cell data and since splicing is mostly binary in single cells (i.e. you have mostly isoform A and not both isoform A and B, citation: <a href="http://www.ncbi.nlm.nih.gov/pubmed/23685454">Shalek et al, <em>Nature</em> (2013)</a>), it becomes a pain to make these annotations, because you have to merge <span class="caps">ALL</span> the files together to create like a terrabyte-sized alignment file (usually a <a href="http://genome.sph.umich.edu/wiki/BAM"><code>*.bam</code> file</a>) and then you are introducing all kinds of craziness because what could have been noise in a few cells gets added up together and becomes&nbsp;signal.</p>
<p>Sigh.</p>
<p>So I&#8217;d much rather use something that quantifies splicing of all junctions, arbitrarily. But when I tried using <code>sjcount</code> I couldn&#8217;t figure out what one of the outputs was (like what&#8217;s the difference between a &#8220;splice junction&#8221; and a &#8220;splice boundary?&#8221;). Plus I realized that the &#8220;splice junction counts&#8221; (<code>*.ssj</code>) file was really similar to the &#8220;<code>SJ.out.tab</code>&#8221; file created by the <a href="https://code.google.com/p/rna-star/"><span class="caps">RNA</span>-<span class="caps">STAR</span></a>&nbsp;aligner.</p>
<p><code>.ssj</code> files look like (I added the column&nbsp;names):</p>
<div class="highlight"><pre><span class="n">chrom</span>   <span class="n">start</span>   <span class="n">stop</span>    <span class="n">strand</span>  <span class="n">overhang</span>    <span class="n">count</span>
<span class="n">chr1</span>    <span class="mi">146509</span>  <span class="mi">155767</span>  <span class="p">.</span>       <span class="mi">20</span>      <span class="mi">1</span>
<span class="n">chr1</span>    <span class="mi">146509</span>  <span class="mi">155767</span>  <span class="p">.</span>       <span class="mi">34</span>      <span class="mi">1</span>
<span class="n">chr1</span>    <span class="mi">146509</span>  <span class="mi">155767</span>  <span class="p">.</span>       <span class="mi">49</span>      <span class="mi">3</span>
<span class="n">chr1</span>    <span class="mi">168165</span>  <span class="mi">169049</span>  <span class="p">.</span>       <span class="mi">10</span>      <span class="mi">2</span>
<span class="n">chr1</span>    <span class="mi">168165</span>  <span class="mi">169049</span>  <span class="p">.</span>       <span class="mi">49</span>      <span class="mi">3</span>
<span class="n">chr1</span>    <span class="mi">694503</span>  <span class="mi">700103</span>  <span class="p">.</span>       <span class="mi">29</span>      <span class="mi">1</span>
<span class="n">chr1</span>    <span class="mi">694503</span>  <span class="mi">700103</span>  <span class="p">.</span>       <span class="mi">41</span>      <span class="mi">1</span>
<span class="n">chr1</span>    <span class="mi">694503</span>  <span class="mi">700103</span>  <span class="p">.</span>       <span class="mi">49</span>      <span class="mi">1</span>
<span class="n">chr1</span>    <span class="mi">894461</span>  <span class="mi">894595</span>  <span class="p">.</span>       <span class="mi">7</span>       <span class="mi">2</span>
<span class="n">chr1</span>    <span class="mi">894461</span>  <span class="mi">894595</span>  <span class="p">.</span>       <span class="mi">9</span>       <span class="mi">5</span>
</pre></div>


<p><span class="dquo">&#8220;</span>overhang&#8221; is the number of base pairs of the sequencing read that overlaps with the upstream exon of the splicing event. The rest of the read is on the downstream exon. In our case, our reads are 100 base pairs (&#8220;100bp&#8221;) so if there&#8217;s 20bp overhanging on one exon, the other 80bp are on the other&nbsp;exon.</p>
<p>While <code>SJ.out.tab</code> files are (again I added column&nbsp;names):</p>
<table border="1" class="dataframe">\n  <thead>\n    <tr style="text-align: right;">\n      <th></th>\n      <th>chrom</th>\n      <th>first_bp_intron</th>\n      <th>last_bp_intron</th>\n      <th>strand</th>\n      <th>intron_motif</th>\n      <th>annotated</th>\n      <th>unique_junction_reads</th>\n      <th>multimap_junction_reads</th>\n      <th>max_overhang</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td> chr1</td>\n      <td> 135332</td>\n      <td> 138297</td>\n      <td> 1</td>\n      <td> <span class="caps">GC</span>/<span class="caps">AG</span></td>\n      <td> False</td>\n      <td> 0</td>\n      <td> 1</td>\n      <td> 38</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td> chr1</td>\n      <td> 146510</td>\n      <td> 155766</td>\n      <td> 2</td>\n      <td> <span class="caps">CT</span>/<span class="caps">AC</span></td>\n      <td>  True</td>\n      <td> 5</td>\n      <td> 1</td>\n      <td> 34</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td> chr1</td>\n      <td> 156895</td>\n      <td> 158300</td>\n      <td> 1</td>\n      <td> <span class="caps">GC</span>/<span class="caps">AG</span></td>\n      <td> False</td>\n      <td> 0</td>\n      <td> 2</td>\n      <td> 17</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td> chr1</td>\n      <td> 168166</td>\n      <td> 169048</td>\n      <td> 2</td>\n      <td> <span class="caps">CT</span>/<span class="caps">AC</span></td>\n      <td>  True</td>\n      <td> 5</td>\n      <td> 3</td>\n      <td> 46</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td> chr1</td>\n      <td> 169265</td>\n      <td> 172556</td>\n      <td> 2</td>\n      <td> <span class="caps">CT</span>/<span class="caps">AC</span></td>\n      <td>  True</td>\n      <td> 0</td>\n      <td> 2</td>\n      <td> 13</td>\n    </tr>\n  </tbody>\n</table>

<p>The start/stop are off by one here because <code>.ssj</code><span class="quo">&#8216;</span>s index is from the exon end/exon start, and this is from the intron start/intron end. So the <span class="caps">SJ</span>.out.tab coordinates are (start from <code>.ssj</code>+1) and (end from <code>.ssj</code>-1).</p>
<p><span class="caps">FYI</span>, the <code>.ssc</code> files looked&nbsp;like:</p>
<div class="highlight"><pre><span class="n">chr1</span>    <span class="mi">894595</span>  <span class="mi">894595</span>  <span class="p">.</span>       <span class="mi">8</span>       <span class="mi">1</span>
<span class="n">chr1</span>    <span class="mi">1228468</span> <span class="mi">1228468</span> <span class="p">.</span>       <span class="mi">1</span>       <span class="mi">1</span>
<span class="n">chr1</span>    <span class="mi">1228468</span> <span class="mi">1228468</span> <span class="p">.</span>       <span class="mi">3</span>       <span class="mi">3</span>
<span class="n">chr1</span>    <span class="mi">1228468</span> <span class="mi">1228468</span> <span class="p">.</span>       <span class="mi">5</span>       <span class="mi">1</span>
<span class="n">chr1</span>    <span class="mi">1228468</span> <span class="mi">1228468</span> <span class="p">.</span>       <span class="mi">6</span>       <span class="mi">1</span>
<span class="n">chr1</span>    <span class="mi">1228468</span> <span class="mi">1228468</span> <span class="p">.</span>       <span class="mi">7</span>       <span class="mi">2</span>
<span class="n">chr1</span>    <span class="mi">1228468</span> <span class="mi">1228468</span> <span class="p">.</span>       <span class="mi">12</span>      <span class="mi">1</span>
<span class="n">chr1</span>    <span class="mi">1228468</span> <span class="mi">1228468</span> <span class="p">.</span>       <span class="mi">13</span>      <span class="mi">2</span>
<span class="n">chr1</span>    <span class="mi">1228468</span> <span class="mi">1228468</span> <span class="p">.</span>       <span class="mi">16</span>      <span class="mi">2</span>
<span class="n">chr1</span>    <span class="mi">1228468</span> <span class="mi">1228468</span> <span class="p">.</span>       <span class="mi">27</span>      <span class="mi">1</span>
</pre></div>


<p>This claims to have the same columns as <code>.ssj</code> files except instead of start/stop it&#8217;s position/position. I was very confused by how you can have a count at a position. Why does <code>chr1:894595</code> have one line while <code>chr1:1228468</code> have a bunch? If you search for <code>894595</code> in the <code>.ssj</code> file, you&nbsp;get:</p>
<div class="highlight"><pre><span class="n">chr1</span>    <span class="mi">894461</span>  <span class="mi">894595</span>  <span class="p">.</span>       <span class="mi">7</span>       <span class="mi">2</span>
<span class="n">chr1</span>    <span class="mi">894461</span>  <span class="mi">894595</span>  <span class="p">.</span>       <span class="mi">9</span>       <span class="mi">5</span>
<span class="n">chr1</span>    <span class="mi">894461</span>  <span class="mi">894595</span>  <span class="p">.</span>       <span class="mi">10</span>      <span class="mi">2</span>
<span class="n">chr1</span>    <span class="mi">894461</span>  <span class="mi">894595</span>  <span class="p">.</span>       <span class="mi">11</span>      <span class="mi">1</span>
<span class="n">chr1</span>    <span class="mi">894461</span>  <span class="mi">894595</span>  <span class="p">.</span>       <span class="mi">12</span>      <span class="mi">1</span>
<span class="n">chr1</span>    <span class="mi">894461</span>  <span class="mi">894595</span>  <span class="p">.</span>       <span class="mi">14</span>      <span class="mi">1</span>
<span class="n">chr1</span>    <span class="mi">894461</span>  <span class="mi">894595</span>  <span class="p">.</span>       <span class="mi">16</span>      <span class="mi">2</span>
<span class="n">chr1</span>    <span class="mi">894461</span>  <span class="mi">894595</span>  <span class="p">.</span>       <span class="mi">19</span>      <span class="mi">1</span>
<span class="n">chr1</span>    <span class="mi">894461</span>  <span class="mi">894595</span>  <span class="p">.</span>       <span class="mi">20</span>      <span class="mi">1</span>
<span class="n">chr1</span>    <span class="mi">894461</span>  <span class="mi">894595</span>  <span class="p">.</span>       <span class="mi">21</span>      <span class="mi">13</span>
<span class="n">chr1</span>    <span class="mi">894461</span>  <span class="mi">894595</span>  <span class="p">.</span>       <span class="mi">22</span>      <span class="mi">1</span>
<span class="n">chr1</span>    <span class="mi">894461</span>  <span class="mi">894595</span>  <span class="p">.</span>       <span class="mi">42</span>      <span class="mi">2</span>
<span class="n">chr1</span>    <span class="mi">894461</span>  <span class="mi">894595</span>  <span class="p">.</span>       <span class="mi">49</span>      <span class="mi">14</span>
</pre></div>


<p>How did all these counts in the last column (which sum to 46 by my calculation) become <strong>1</strong> in the <code>.ssc</code> file? I was pretty&nbsp;lost.</p>
<p><span class="caps">ANYWAYS</span> all of this is a buildup to say that I used Pervouchine et al&#8217;s method of quantifying alternative junctions as asking, when this upstream exon (splicing <em>donor</em>, <em>D</em> in figure below) is used with this downstream exon (splicing <em>acceptor</em>, <em>A</em> in figure below), how often does that happen relative to all other donors and acceptors? In the figure below, the splicing event of interest is a thick, bold line, joined by a solid arc representing the spliced read. The dotted lines represent that this donor could be spliced to a bunch of different acceptor&nbsp;sites. </p>
<p><img alt="Donor with alternative acceptors on left, Acceptor with alternative donors on right" src="http://bioinformatics.oxfordjournals.org/content/29/2/273/F2.medium.gif" /></p>
<p>This is quantified with a $\Psi_5$ (&#8220;Percent spliced-in&#8221;, or &#8220;Psi&#8221; of the donor, which is at the 5&#8217; end of the <span class="caps">RNA</span>) and $\Psi_3$ (&#8220;Psi&#8221; of the acceptor, located at the 3&#8217; end of the <span class="caps">RNA</span>)&nbsp;scores:</p>
<p><img alt="" src="http://bioinformatics.oxfordjournals.org/content/29/2/273/embed/graphic-3.gif" /></p>
<p>Where the summation is over all other possible acceptors (in the case of <em>A&#8217;</em>) or all other possible donors (in the case of <em>D&#8217;</em>).</p>
<p>So using the <span class="caps">SJ</span>.out.tab files from <span class="caps">RNA</span>-<span class="caps">STAR</span> and the magic of <a href="http://pandas.pydata.org/"><code>pandas</code></a>, and a <a href="https://github.com/olgabot/sj2psi/blob/master/sj2psi/__init__.py#L64">bit of filtering</a>, you can calculate these scores very&nbsp;easily:</p>
<div class="highlight"><pre><span class="n">sj</span><span class="p">[</span><span class="err">&#39;</span><span class="n">psi5</span><span class="err">&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">sj</span><span class="p">.</span><span class="n">groupby</span><span class="p">([</span><span class="err">&#39;</span><span class="n">chrom</span><span class="err">&#39;</span><span class="p">,</span> 
    <span class="err">&#39;</span><span class="n">first_bp_intron</span><span class="err">&#39;</span><span class="p">])[</span><span class="err">&#39;</span><span class="n">total_filtered_reads</span><span class="err">&#39;</span><span class="p">].</span><span class="n">transform</span><span class="p">(</span><span class="n">lambda</span> <span class="n">x</span><span class="o">:</span> <span class="n">x</span><span class="o">/</span><span class="n">x</span><span class="p">.</span><span class="n">sum</span><span class="p">())</span>
</pre></div>


<p>While this does $\Psi$ scores from raw counts which is frequentist and I prefer to be Bayesian, this is a reasonable&nbsp;start.</p>
<p>Check out the package I wrote for this: https://github.com/olgabot/sj2psi I&#8217;d love your feedback and contribution. It&#8217;s also <a href="https://pypi.python.org/pypi/sj2psi/0.0.1">available on PyPI</a>, so you can <code>pip install sj2psi</code> and start using it today! It&#8217;s under the <span class="caps">MIT</span> license so just attribute me&nbsp;:)</p>
            <a href="#" class="go-top">Go Top</a>
    <div class="comments">
        <div id="disqus_thread"></div>
        <script type="text/javascript">
            /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
            var disqus_shortname = "sciencemeetproductivity"; // required: replace example with your forum shortname

            /* * * DON'T EDIT BELOW THIS LINE * * */
            (function() {
                var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
                dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
                (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
            })();
        </script>
        <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
        <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
    </div>
<footer class="footer">
    <p>&copy; Olga Botvinnik &ndash;
        Built with <a href="https://github.com/PurePelicanTheme/pure-single">Pure Theme</a>
        for <a href="http://blog.getpelican.com/">Pelican</a>
    </p>
</footer>        </div>
    </div>
    </div>
    <script>
        var $top = $('.go-top');

        // Show or hide the sticky footer button
        $(window).scroll(function() {
            if ($(this).scrollTop() > 200) {
                $top.fadeIn(200);
            } else {
                $top.fadeOut(200);
            }
        });

        // Animate the scroll to top
        $top.click(function(event) {
            event.preventDefault();
            $('html, body').animate({scrollTop: 0}, 300);
        })

        // Makes sure that the href="#" attached to the <a> elements
        // don't scroll you back up the page.
        $('body').on('click', 'a[href="#"]', function(event) {
            event.preventDefault();
        });
    </script>

</body>
</html>